hellabot.go

```diff
@@ -310,13 +316,31 @@ func (bot *Bot) Close() error {
 	return nil
 }
 
-// AddTrigger adds a given trigger to the bots handlers
+// AddTrigger adds a given trigger to the bot's handlers
 func (bot *Bot) AddTrigger(t Trigger) {
+	bot.triggersMu.Lock()
+	defer bot.triggersMu.Unlock()
 	bot.triggers = append(bot.triggers, t)
 }
 
+// DropTrigger removes a trigger from the bot's handlers
+func (bot *Bot) DropTrigger(t Trigger) bool {
+	bot.triggersMu.Lock()
+	defer bot.triggersMu.Unlock()
+	for i, tt := range bot.triggers {
+		if t.Name != "" && t.Name == tt.Name {
+			bot.triggers = append(bot.triggers[:i], bot.triggers[i+1:]...)
```

Comment:
Okay this one I don't follow. How will the copy be mutated if we have a mutex when making the copy?

Reply:
A slice is a [struct with three fields](https://golang.org/pkg/reflect/#SliceHeader). The length is how many elements are in the slice, and the capacity is the number of elements that can fit into the underlying data array. When you append to the slice, it checks to see if there is enough capacity for the additional element. If there isn't enough room, a new array is allocated and the elements are copied to that. The slice expression `bot.triggers[:i]` will create a new slice that points to the same underlying array. The `triggers` copy we make also points to the same underlying array. By changing the slice expression to `bot.triggers[:i:i]` we are artificially setting the capacity which will force it to create a new underlying data array.

Reply:
If the new `triggers` slice is the concern, shouldn't the new array/safe copy happen there instead of here?

Reply:
@bbriggs it could, but I figured this would be a better place since it happens way less often.

Reply:
That's fair. Copying a new slice for every single message can get really busy. I'm just thinking through ways it can be both _safe_ and _obvious_ what's happening (and in the right location)

Reply:
I think the performance impact is negligible. Copying each time we execute is fine with me.